### PR TITLE
feat: #26 add reactive authorization

### PR DESCRIPTION
**Pull request by Chlenix**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -25,6 +25,37 @@
...
         </dependency>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-security</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.security.oauth.boot</groupId>
+            <artifactId>spring-security-oauth2-autoconfigure</artifactId>
+            <version>${parent.version}</version>
+        </dependency>
```
Comment (azaxarov) [inline pom.xml]: Is it still in use?

File: src/main/java/no/fdk/dataservicecatalog/config/SecurityConfig.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import org.springframework.http.HttpStatus;
+import org.springframework.security.config.annotation.method.configuration.EnableReactiveMethodSecurity;
+import org.springframework.security.config.annotation.web.reactive.EnableWebFluxSecurity;
+import org.springframework.security.config.web.server.ServerHttpSecurity;
+import org.springframework.security.web.server.SecurityWebFilterChain;
+import reactor.core.publisher.Mono;
+
+@Configuration
+@RequiredArgsConstructor
+
```
Comment (azaxarov) [inline src/main/java/no/fdk/dataservicecatalog/config/SecurityConfig.java]: Feeling rebellious?

File: src/main/java/no/fdk/dataservicecatalog/config/SecurityProperties.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package no.fdk.dataservicecatalog.config;
+
+
+import lombok.Data;
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.stereotype.Component;
+
+@Data
+@Component
+@ConfigurationProperties("security.jwks")
```
Comment (azaxarov) [inline src/main/java/no/fdk/dataservicecatalog/config/SecurityProperties.java]: use security only and jwksUri as a variable

File: src/main/java/no/fdk/dataservicecatalog/security/AuthenticationManager.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+public class AuthenticationManager implements ReactiveAuthenticationManager {
+
+    private final SecurityProperties securityProperties;
+    private AuthoritiesExtractor authoritiesExtractor = new FixedAuthoritiesExtractor();
+
+    @Override
+    public Mono<Authentication> authenticate(Authentication authentication) {
+        var token = authentication.getCredentials();
+        return NimbusReactiveJwtDecoder.withJwkSetUri(securityProperties.getUri())
+                .build().decode(token.toString())
```
Comment (azaxarov) [inline src/main/java/no/fdk/dataservicecatalog/security/AuthenticationManager.java]: not sure .toString() is needed here

File: src/main/java/no/fdk/dataservicecatalog/security/AuthenticationManager.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+@Component
+@RequiredArgsConstructor
+public class AuthenticationManager implements ReactiveAuthenticationManager {
+
+    private final SecurityProperties securityProperties;
+    private AuthoritiesExtractor authoritiesExtractor = new FixedAuthoritiesExtractor();
+
+    @Override
+    public Mono<Authentication> authenticate(Authentication authentication) {
+        var token = authentication.getCredentials();
```
Comment (azaxarov) [inline src/main/java/no/fdk/dataservicecatalog/security/AuthenticationManager.java]: remove token variable and use .getCredentials() where needed

File: src/main/java/no/fdk/dataservicecatalog/security/SecurityContextRepository.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+            authToken = authHeader.replace(TOKEN_PREFIX, "").trim();
+        } else {
+            log.warn("couldn't find bearer string, will ignore the header.");
+        }
+        if (authToken != null) {
+            Authentication auth = new UsernamePasswordAuthenticationToken(authToken, authToken);
+            return this.authenticationManager.authenticate(auth).map(SecurityContextImpl::new);
+        } else {
+            return Mono.empty();
+        }
```
Comment (azaxarov) [inline src/main/java/no/fdk/dataservicecatalog/security/SecurityContextRepository.java]: This logic needs cleanup - very difficult to follow what's going on here

File: src/main/java/no/fdk/dataservicecatalog/service/DataServiceRegistrationService.java [hunk 7]
```diff
@@ -74,6 +74,7 @@ private DataService applyPatch(DataService dataService, Map<String, Object> part
         });
     }
 
+    //    @PreAuthorize("@permissionService.hasPermission(#catalogId, 'organization', 'read')")
```
Comment (azaxarov) [inline src/main/java/no/fdk/dataservicecatalog/service/DataServiceRegistrationService.java]: Delete?

File: src/main/resources/application.yaml [hunk 8]
```diff
@@ -19,8 +19,16 @@ spring:
     template:
       routing-key: 'dataservice.publisher.HarvestTrigger'
       exchange: 'harvests'
+
+security:
+  jwks:
+    uri: ${SSO_HOST}/auth/realms/fdk/protocol/openid-connect/certs
```
Comment (azaxarov) [inline src/main/resources/application.yaml]: rename to jwks-uri

File: src/main/resources/application.yaml [hunk 9]
```diff
@@ -19,8 +19,16 @@ spring:
...
 ---
 
+security:
+  jwks:
+    uri: https://sso.staging.fellesdatakatalog.digdir.no/auth/realms/fdk/protocol/openid-connect/certs
```
Comment (azaxarov) [inline src/main/resources/application.yaml]: same here

File: pom.xml (hunk 1 above)
Comment (Chlenix) [reply pom.xml]: AuthorityExtractor

**Pull request merged**